### PR TITLE
glTF: support WIP mesh_compression_open3dgc extension

### PR DESCRIPTION
**Pull request by kainino0x**
(Do not merge; extension spec may change. KhronosGroup/glTF#398)

* [x] Is ThirdParty handled correctly?
* [ ] Wait for finalized extension spec
* [ ] What's wrong with binary-encoded Open3DGC (as opposed to ascii-encoded)? It isn't loading (last I checked).
* [ ] Add tests
* [x] Try multiple primitives per mesh, once the converter supports that
* [x] Try integer attributes
* [ ] Any TODOs/FIXMEs
* [ ] Remove commented-out timing code (`var timer =`, etc.)
* [ ] Look into the performance of the NYC dataset - is the problem something we can address, or is this compression algorithm just not good for this dataset? (Probably the latter.)

* [ ] Performance comparison with/without mesh compression (both filesize/network and decompression/CPU)
* [ ] Find threshold for when to use web workers
* [ ] Web worker slow on IE and probably some other platforms (no transferable object support according to MDN).
* [x] Look into [`SharedArrayBuffer`](https://www.chromestatus.com/feature/4570991992766464), currently unstandardized and only supported in Firefox Nightly
* [ ] Look into using some kind of thread pool (e.g. work stealing or work queue), possibly along the lines of threads.js. (For efficiently parallelizing loading of many glTF "primitives" - currently jobs are scheduled round-robin, which can result in very bad balancing.)
* [ ] ?

### CODE REVIEW

**Comment by pjcozzi** [comment]
> Is ThirdParty handled correctly?

Yes, but also add it to [LICENSE.md](https://github.com/AnalyticalGraphicsInc/cesium/blob/master/LICENSE.md)

**Comment by pjcozzi** [comment]
Update CHANGES.md.

**Comment by pjcozzi** [comment]
Does compress work for models with animations and skins (even if it only compresses geometry?).  We should ship our default models compressed if possible.

**Comment by pjcozzi** [comment]
Captain obvious here reminding you to link to this issue in a new comment in https://github.com/KhronosGroup/glTF/issues/398 so the glTF folks can see what a loader looks like.

**Comment by pjcozzi** [comment]
Once you have some timing numbers, let us know if you think we should do decompression in a web worker.

File: Source/Scene/Model.js [hunk 1]
```diff
@@ -1267,6 +1292,79 @@ define([
...
+            return;
+        }
+
+        var extensions = model.gltf.extensions;
+        if (!defined(extensions) || !defined(extensions.mesh_compression_open3dgc)) {
+            return;
+        }
+
+        var decompressedBuffers = model.gltf.extensions.mesh_compression_open3dgc.decompressedBuffers;
+        while (loadResources.decompressedBuffersToCreate.length > 0) {
```
Comment (pjcozzi) [inline Source/Scene/Model.js]: It's not our long-term throttling solution, but this should probably check `model.asynchronous` like [this](https://github.com/kainino0x/cesium/blob/mesh_compression_open3dgc/Source/Scene/Model.js#L1468) **unless** decompression is very fast.

File: Source/Scene/Model.js [hunk 2]
```diff
@@ -1267,6 +1292,79 @@ define([
...
+            }
+            var numFloatAttributes = ifs.GetNumFloatAttributes();
+            for (var i = 0; i < numFloatAttributes; i++) {
+                if (ifs.GetNFloatAttribute(i) > 0) {
+                    len = ifs.GetFloatAttributeDim(i) * ifs.GetNFloatAttribute(i);
+                    ifs.SetFloatAttribute(i, new Float32Array(decompressedArrayBuffer, decompressedBufferOffset, len));
+                    decompressedBufferOffset += len * Float32Array.BYTES_PER_ELEMENT;
+                }
+            }
+            // TODO: handle int attributes?
```
Comment (pjcozzi) [inline Source/Scene/Model.js]: Perhaps add this to the TODO list in the pull request since we generally don't merge pull requests with TODO comments.

**Comment by pjcozzi** [comment]
Looks good - glad this was straightforward.  ~1 day?  That would be good messaging.

**Comment by kainino0x** [comment]
Well, I did only spend about 4 hours on it (from start to last commit). After that, the work I did was on fixing the converter and switching to ascii compression.

But I have not tried to do anything with animations, so that is not really a complete number

**Comment by pjcozzi** [comment]
@kainino0x this breaks the build: https://travis-ci.org/AnalyticalGraphicsInc/cesium/builds/75361956

**Comment by jbo023** [comment]
Ho,
i am working with the 3D-TilesBranch. Unfortunatly the line 
https://github.com/kainino0x/cesium/blob/mesh_compression_open3dgc/Source/ThirdParty/o3dgc.js#L188
breaks cesium on ios, because the global performance object is not working on ios. 
A further check for performance should fix the problem. 

 if ((typeof performance !== 'undefined') && performance.timing && performance.timing.navigationStart){

**Comment by pjcozzi** [comment]
I think that will be fixed in this pull request: https://github.com/kainino0x/cesium/pull/1/files#diff-405ef91bc51513b45be451d758ea4773L184

**Comment by mramato** [comment]
@jbo023 The head of the [3d-tiles](https://github.com/AnalyticalGraphicsInc/cesium/tree/3d-tiles) is already fixed and has most of the code in this PR.  The line you linked to doesn't exist in that branch.

**Comment by jbo023** [comment]
Thanks, 
i was still on the commit from the 10.08, i must have missed the new commit.

**Comment by kainino0x** [comment]
Returning to work on this over winter break. A few days ago, I merged master into the compression branch. However, I haven't verified the correctness of the merge. Right now I'm stuck on a strange issue: running my test sandbox pages crashes Chrome (both stable and canary). It doesn't show the model in Firefox right now, either. A merge error seems likely, but I'm trying to track down what causes the crash. @pjcozzi

**Comment by kainino0x** [comment]
The line seems to be o3dgc.js:1248
`tmp_data = new Int32Array(this.m_allocated);`

Apparently the web worker is having an issue with memory allocation, now?

**Comment by pjcozzi** [comment]
Does this work OK without web workers?

**Comment by kainino0x** [comment]
Turns out all I did was mangle my test setup - it's trying to allocate a 3 GB float array because I fed it a gzipped bin as a regular bin. I've got things working again.

**Comment by pjcozzi** [comment]
Nice :grinning:

**Comment by kainino0x** [comment]
Notes from today:
* Preliminary results show decompression takes about the same amount of time with or without workers (though the variance in both cases is high), except on the first few loads when web workers are getting created (takes an extra few hundred ms per web worker). This is based on small models.
* For models with many parts, performance is improved using web workers, due to parallelism. A simple test (loading a 7-part dragon with 7 threads) shows a pretty good improvement: roughly 900ms unthreaded, roughly 400ms threaded. This _should_ also apply to loading multiple models at once. However, good performance may require a better scheduling algorithm than round-robin (e.g. a queue or work-stealing; whatever works with `TaskProcessor`/web workers).
* TODO: just for easier performance testing, find a hack to pre-allocate the web workers (possibly by just loading a model with many parts).
* TODO: finish other merge so I can convert more models to gltf.

**Comment by pjcozzi** [comment]
Thanks for the update.

> A simple test (loading a 7-part dragon with 7 threads) shows a pretty good improvement: roughly 900ms unthreaded, roughly 400ms threaded.

Did you try with only 1 or 2 web workers?

> However, good performance may require a better scheduling algorithm than round-robin (e.g. a queue or work-stealing; whatever works with TaskProcessor/web workers).

Agreed.  There's some initial work here, but it still for the main thread only: [JobScheduler](https://github.com/AnalyticalGraphicsInc/cesium/blob/3d-tiles/Source/Scene/JobScheduler.js).

**Comment by kainino0x** [comment]
I haven't tried with fewer workers yet.

I tried searching for a work queue or thread pool; they do exist:
https://github.com/andywer/threads.js

**Comment by mramato** [comment]
For what it's worth, The NYC data set performs noticeably worse with compression than without.  Even though the payload is much smaller with compression, we only ever download it once and decompression caused noticeable lag when retrieving from the cache.  I'm not saying it's not a fixable problem, but we should definitely test it out again before merging this to see where potential problems are.  It's also possible that large simple models is simply not an ideal use case for using compression and this feature better serves other cases.

**Comment by kainino0x** [comment]
I've been thinking about that. Thanks for reminding me, I should put it in the checklist.

I do think that the NYC dataset is not generally conducive to this compression algorithm, though.

**Comment by pjcozzi** [comment]
We are likely to use [WEB3D_quantized_attributes](https://github.com/KhronosGroup/glTF/blob/master/extensions/Vendor/WEB3D_quantized_attributes/README.md) - where decompression is free - for datasets like NYC - and probably most 3D tilesets.  This compression will mostly serve individual assets.

**Comment by pjcozzi** [comment]
This branch is also in the main Cesium repo now: https://github.com/AnalyticalGraphicsInc/cesium/compare/mesh-compression-open3dgc

**Comment by mramato** [comment]
Should we just close this (we can keep the branch around if needed)?

**Comment by pjcozzi** [comment]
OK to close, but we'll keep the branch for when someone picks up https://github.com/KhronosGroup/glTF/issues/398.

**Pull request closed without merging**